File: app/src/main/java/club/gitmad/chat/MainActivity.java
package club.gitmad.chat;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

public class MainActivity extends AppCompatActivity {

    private static final String userId = UUID.randomUUID().toString();

    private RecyclerView rvMessages;
    private MessageAdapter messageAdapter;
    private List<Message> messages;

    private EditText etMessage;
    private Button btnSend;

    private Message currMessage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        rvMessages = findViewById(R.id.rvMessages);
        etMessage = findViewById(R.id.etMessage);
        btnSend = findViewById(R.id.btnSend);

        messages = new ArrayList<>();
        messageAdapter = new MessageAdapter(messages);
        rvMessages.setLayoutManager(new LinearLayoutManager(this));
        rvMessages.setAdapter(messageAdapter);

        FirebaseFirestore.getInstance().collection("messages")
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
                    @Override
                    public void onEvent(@Nullable QuerySnapshot queryDocumentSnapshots, @Nullable FirebaseFirestoreException e) {
                        if (e != null) {
                            Log.e("MainActivity", "Firebase Cloud Firestore - Initial listener", e);
                            return;
                        }

                        for (QueryDocumentSnapshot doc : queryDocumentSnapshots) {
                            Message newMessage = doc.toObject(Message.class);
                            if (!messages.contains(newMessage)) {
                                messages.add(newMessage);
                            }
                        }

                        Collections.sort(messages, new Comparator<Message>() {
                            @Override
                            public int compare(Message o1, Message o2) {
                                return Long.compare(o1.getTime(), o2.getTime());
                            }
                        });

                        messageAdapter.notifyDataSetChanged();
                        rvMessages.smoothScrollToPosition(messages.size());
                    }
                });
    }
}

File: app/src/main/java/club/gitmad/chat/MainActivity.kt
package club.gitmad.chat

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.firebase.ui.auth.AuthUI
import com.google.firebase.auth.ktx.auth
import com.google.firebase.firestore.EventListener
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.firestore.ktx.toObject
import com.google.firebase.ktx.Firebase
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : AppCompatActivity() {

    private val RC_SIGN_IN = 123

    private var userId = ""
    private var userName = ""

    private lateinit var messageAdapter: MessageAdapter
    private var messages: MutableList<Message> = mutableListOf()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        signIn()

        messageAdapter = MessageAdapter(messages)
        rvMessages.adapter = messageAdapter

        Firebase.firestore.collection("messages")
                .addSnapshotListener(EventListener { queryDocumentSnapshots, e ->
                    if (e != null) {
                        Log.e("MainActivity", "Firebase Cloud Firestore - Initial listener", e)
                        return@EventListener
                    }

                    queryDocumentSnapshots?.forEach {
                        val newMessage = it.toObject<Message>()
                        if (!messages.contains(newMessage)) {
                            messages.add(newMessage)
                        }
                    }

                    messages.sortWith { o1, o2 -> o1.time.compareTo(o2.time) }

                    messageAdapter.notifyDataSetChanged()
                    rvMessages.smoothScrollToPosition(messages.size)
                })

        btnSend.setOnClickListener {
            // TODO 1. Get the text of etMessage

            // TODO 2. Create a Message object
            val message = Message(userId, userName)

            // TODO 3. Write to the FirebaseFirestore collection called "messages"
            // https://firebase.google.com/docs/firestore/manage-data/add-data#add_a_document
        }
    }

    private fun signIn() {
        // TODO: If user is logged in, set the userId and userName. Else, log in.
        // https://firebase.google.com/docs/auth/android/firebaseui#kotlin+ktx
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == RC_SIGN_IN) {
            if (resultCode == Activity.RESULT_OK) {
                // TODO: Get the userId and userName and save them to the fields
                // https://firebase.google.com/docs/auth/android/firebaseui#kotlin+ktx_1
            }
        }
    }
}
